src/Microsoft.AspNetCore.Authentication.Core/AuthenticationSchemeProvider.cs

```diff
@@ -93,43 +90,49 @@ public Task<AuthenticationScheme> GetDefaultForbidSchemeAsync()
         /// <summary>
         /// Returns the scheme that will be used by default for <see cref="IAuthenticationService.SignInAsync(HttpContext, string, System.Security.Claims.ClaimsPrincipal, AuthenticationProperties)"/>.
         /// This is typically specified via <see cref="AuthenticationOptions.DefaultSignInScheme"/>.
-        /// Otherwise, if only a single scheme exists, that will be used, if more than one exists, null will be returned.
+        /// If only a single sign in handler scheme exists, that will be used, if more than one exists,
+        /// this will fallback to <see cref="GetDefaultAuthenticateSchemeAsync"/>.
         /// </summary>
         /// <returns>The scheme that will be used by default for <see cref="IAuthenticationService.SignInAsync(HttpContext, string, System.Security.Claims.ClaimsPrincipal, AuthenticationProperties)"/>.</returns>
-        public Task<AuthenticationScheme> GetDefaultSignInSchemeAsync()
+        public virtual async Task<AuthenticationScheme> GetDefaultSignInSchemeAsync()
         {
             if (_options.DefaultSignInScheme != null)
             {
-                return GetSchemeAsync(_options.DefaultSignInScheme);
+                return EnsureAssignable<IAuthenticationSignInHandler>(await GetSchemeAsync(_options.DefaultSignInScheme));
             }
-            if (_map.Count == 1)
+            if (_signInHandlers.Count == 1)
             {
-                return Task.FromResult(_map.Values.First());
+                return _signInHandlers[0];
             }
-            return Task.FromResult<AuthenticationScheme>(null);
+            return EnsureAssignable<IAuthenticationSignInHandler>(await GetDefaultAuthenticateSchemeAsync());
         }
 
         /// <summary>
         /// Returns the scheme that will be used by default for <see cref="IAuthenticationService.SignOutAsync(HttpContext, string, AuthenticationProperties)"/>.
         /// This is typically specified via <see cref="AuthenticationOptions.DefaultSignOutScheme"/>.
-        /// Otherwise, this will fallback to <see cref="GetDefaultSignInSchemeAsync"/> .
+        /// If only a single sign out handler scheme exists, that will be used, if more than one exists,
+        /// this will fallback to <see cref="GetDefaultSignInSchemeAsync"/> if that supoorts sign out.
         /// </summary>
         /// <returns>The scheme that will be used by default for <see cref="IAuthenticationService.SignOutAsync(HttpContext, string, AuthenticationProperties)"/>.</returns>
-        public Task<AuthenticationScheme> GetDefaultSignOutSchemeAsync()
+        public virtual async Task<AuthenticationScheme> GetDefaultSignOutSchemeAsync()
         {
             if (_options.DefaultSignOutScheme != null)
             {
-                return GetSchemeAsync(_options.DefaultSignOutScheme);
+                return EnsureAssignable<IAuthenticationSignOutHandler>(await GetSchemeAsync(_options.DefaultSignOutScheme));
```

Comment:
Oh right, ok I can back out the extra complexity in the scheme provider then, and leave the type interface enforcement to the service